### DISCUSSION TITLE
[Python] Add `from_numpy` + replacement scan ability for (dictionary of) numpy arrays

### DISCUSSION THREAD

**Discussion by Tishj** [duckdb/duckdb#6245] [Ideas]
We already support creating a relation over a pandas dataframe, and we also support doing a replacement scan over a pandas(+ polars) dataframe, so it makes sense to also extend this to numpy.

Also this *should* be relatively easy to add.

**Comment by Tishj** [comment]
Probably we want to support these 3 ways of creating from numpy:
- Dict[str (name), array] - use the keys of the dict to set the column names
- List[array] - use default names (column0, column1, etc..)
- array - same as List

Probably `TryReplacement` in DuckDBPyConnection will need to be changed as part of the code to support the replacement scan
This commit of a recent PR by Mytherin adding the Polars dataframes could be a reference for this change https://github.com/duckdb/duckdb/pull/6181/commits/7e6f21343b7ac3b0014d15c3bb9155d4993e8425#diff-7c2052a992507edd04315acf93ff1c68628dd402ddd4c0658db600d9af5674aa

**Reply by ghost** [reply]
Hi Tishj,

  For **Dict[str (name), array]** and **List[array]**, are we assuming that the array can only be one dimensional vectors?

  And for **array - same as List**, are we assuming that array can only be a  one dimensional vector or a 2-d matrix?

**Reply by Tishj** [reply]
It's probably best to assert that the arrays are one dimensional for now, or throw a NotImplementedException if the user attempts to supply a multi dimensional array.
The idea is to just map array -> column one to one

**Reply by ghost** [reply]
Yes. I will use asserts and exceptions in the implementation. Thanks
